Mazy Mice

rabestro opened the pull request:
## Add Practice Exercise: Mazy Mice

**Exercise Description:**

This pull request introduces a new practice exercise called "Mazy Mice".

Mickey and Minerva are two clever mice who love to navigate mazes to find cheese. They prefer mazes that have only one correct path to the cheese, with no loops or inaccessible sections.

The goal of this exercise is to write a program that generates "perfect" mazes. A perfect maze is defined as a maze where every cell is reachable and there is exactly one path between any two cells. The generated maze should be rectangular, with an entrance on the left and an exit on the right.

**Key Features of the Exercise:**

* **Maze Generation:** Implement an algorithm (e.g., Depth First Search, Randomized Kruskal's, Randomized Prim's) to generate a perfect maze of given dimensions (rows and columns).
* **Output Formatting:** The maze should be represented as a string using box-drawing characters (e.g., `┌`, `─`, `│`, `└`, `┘`, `├`, `┤`, `┬`, `┴`, `┼`) for walls and passages, and an arrow symbol (`⇨`) for the entrance and exit.
* **Seeded Generation:** The generator should accept an optional seed. If a seed is provided, the same maze must be generated every time for that seed. If no seed is provided, a random maze should be generated.
* **Constraints:** Maze dimensions (rows and columns) should typically be within a specified range (e.g., 5 to 100 cells).

This PR includes a `canonical-data.json` file with test cases covering:

* **Maze Properties:**
    * Correct dimensions (width and height based on input rows/cols).
    * Usage of valid characters for walls, paths, entrance, and exit.
    * Presence of a single entrance and a single exit.
    * Verification of "perfectness" (single path, no loops, no isolated sections) for various maze sizes and aspect ratios.
* **Seeded Generation:**
    * Tests to ensure that providing a specific seed results in a deterministic, predictable maze output. Several cases include the full expected maze string for a given seed.
* **Randomness:**
    * A test case description to guide tracks in verifying that omitting the seed leads to different mazes on subsequent runs (acknowledging this is often a procedural test at the track level).
* **Error Handling:**
    * Test cases for invalid input dimensions (e.g., rows/columns outside the allowed range), expecting specific error messages.

This exercise aims to provide a challenging and engaging problem involving algorithm design, data structure manipulation (representing the maze), and careful output formatting.

Commit:
Add Mazy Mice Exercise to the Repository

This commit introduces a new exercise named 'Mazy Mice'. The exercise is designed to implement a maze generator that produces perfect mazes, i.e., mazes that have only one correct path, without any isolated sections. User must also adhere to the restrictions about the maze's size and layout depicted in the problem description. Tasks also include a description of how the mazes should be visually represented. The commit also includes canonical data for testing the described functionality.

Commit:
Refactor Mazy Mice description format for better readability

Adjusted the format of the 'Mazy Mice' exercise description to enhance legibility. Added separate lines before the maze example sections to clearly differentiate them from the surrounding text. These changes improve the user's reading experience, making it easier for them to understand the task.

rabestro commented:
> Please add a forum link to this discussion.
https://forum.exercism.org/t/add-mazy-mice-exercise-to-the-repository/7113

ErikSchierboom commented:
Thanks for creating the PR! I am not entirely sure how useful the canonical data is right now as the current format does not really allow for easy test generation. This is not something this PR can do a ton about, as randomness is really hard. 

It might be worth looking at how we deal with randomness in other canonical data:

- https://github.com/exercism/problem-specifications/blob/main/exercises/diffie-hellman/canonical-data.json
- https://github.com/exercism/problem-specifications/blob/main/exercises/dnd-character/canonical-data.json

I think the idea of a maze generator is really cool, I'd just like us to think a bit on how to best structure the canonical data to allow it to be as helpful as it can be for implementing tracks. CC @exercism/reviewers

Commit:
Merge branch 'main' into mazy-mice

Commit:
Update test case comments in mazy-mice exercise

Expanded the comment section in the canonical-data.json file for the "mazy-mice" exercise to provide a more detailed guideline on the required checks for a generated maze. These include confirming correct maze dimensions, valid character use, singular entrance and exit, perfection of the maze and its randomness.

Commit:
Correct indentation in maze diagrams

Indentation in the maze diagrams within the mazy-mice exercise was inconsistent. The changes ensure tabs are uniformly used for all lines to improve readability and consistency in presentation.

Commit:
Update section heading levels in mazy-mice description

Headings levels in the "mazy-mice/description.md" were adjusted for better organization. '# Hints' was changed to '## Hints' and '## Maze generation' and '## Box drawing characters' changed to '###'. This allows for an improved logical hierarchy and readability of the document.

Commit:
Corrected the positioning of Box-drawing link in mazy-mice description

Removed the Box-drawing reference link from the code block area to the bottom of the document under the "Maze generation" section. This change was proposed to enhance the readability of the document and provide the link at a more appropriate location.

Commit:
Add additional test cases to Mazy Mice exercise

Expanded test coverage for the 'generateMaze' functionality in the Mazy Mice exercise. This includes tests for maze dimensions, character validity, maze entrance and exit, and more. Additionally, replaced 'createMaze' with 'generateMaze' in existing test cases to ensure consistency. These changes aim to provide a more comprehensive evaluation of the generateMaze function's correctness and robustness.

rabestro commented:
> I think the idea of a maze generator is really cool, I'd just like us to think a bit on how to best structure the canonical data to allow it to be as helpful as it can be for implementing tracks. CC @exercism/reviewers

I have updated the canonical data with additional test cases and provided a Java implementation, which can be found at https://github.com/exercism/java/pull/2355. However, I am encountering an issue with my markdown formatting and the CI/CD process is failing as a result. I am unsure of what steps to take to correct the formatting error.

ErikSchierboom commented:
@rabestro You can format the file using:

```
yarn install
yarn run format-md
```

ErikSchierboom commented:
> I have updated the canonical data with additional test cases

The expected values are still string values. Did you look at the examples I listed? If so, what did you think? CC @exercism/reviewers for thoughts on how to best structure this canonical data

Commit:
Correct casing and formatting in description.md

Changed the casing of a markdown hyperlink reference to maintain consistency. Also adjusted the formatting of the 'Box drawing characters' table for better legibility. These changes aim to improve the readability and adherence of the documentation to markdown standards.

rabestro commented:
> @rabestro You can format the file using:
> 
> ```
> yarn install
> yarn run format-md
> ```

fixed, thank you!

Commit:
Update expected results format in canonical-data.json

In the 'exercises/mazy-mice/canonical-data.json' file, the results format for each test case has been updated. Instead of providing an expected result as a string, a boolean or an object structure is used. The change is performed to improve the test result validation process. Boolean values help to clearly identify the pass/fail status, while the object returns the expected maze dimensions, allowing for more detailed automated test result evaluations.

ErikSchierboom commented:
Did you mean to close it?

rabestro commented:
> > I have updated the canonical data with additional test cases
> 
> The expected values are still string values. Did you look at the examples I listed? If so, what did you think? CC @exercism/reviewers for thoughts on how to best structure this canonical data

I changed to numbers and boolean values

rabestro commented:
> Did you mean to close it?

No, I just close the comments :)

rabestro committed:
Merge branch 'exercism:main' into mazy-mice

Commit:
Update exercise metadata and add validation for maze dimensions

Renamed exercise from "maze" to "mazy-mice" in metadata. Added new validation rule enforcing rows and columns to be within 5 to 100, along with a test case for rows less than the minimum allowed.

senekor commented on exercises/mazy-mice/description.md:
```diff
@@ -0,0 +1,77 @@
+|     ┴     | box drawings light up and horizontal   | U+2534  |
+|     ┘     | box drawings light up and left         | U+2518  |
+|     ├     | box drawings light vertical and right  | U+2520  |
+|     ⇨     | rightwards white arrow                 | U+21E8  |
```
This arrow character seems to be rendered with an inconsistent width. Why not use a monospace-friendly arrow symbol like `>` ?

rabestro replied:
Hi @senekor 

Thanks for your feedback on the Mazy Mice PR, specifically regarding the ⇨ (RIGHTWARDS WHITE ARROW) symbol and its rendering. That's a very valid point – ensuring consistent rendering of characters, especially for visual elements like the maze in the problem description, is important for clarity for all users.

I've been thinking about this in the broader context of how Exercism exercises are implemented across different language tracks, and I wanted to share a few perspectives:

Diverse Implementation Approaches: For many language tracks (e.g., Scala, Java, Python), students likely won't be required to output the maze as a raw string. Instead, the exercise would often involve returning a data structure representing the maze. For instance, a Scala implementation might have a signature like this:

Scala
```scala
// Enum to represent different types of cells in the maze
enum CellType:
  case WALL
  case PATH
  case ENTRANCE
  case EXIT
  // Potentially other types if needed by a specific track's testing approach

// Function to generate the maze
def generateMaze(rows: Int, cols: Int, seed: Option[Int] = None): List[List[CellType]] = {
  // ... maze generation logic returning the structured maze ...
  ??? // Placeholder for actual implementation
}
```

In these cases, the specific graphical symbols used in the description.md serve primarily as a visual illustration of the concepts (like ENTRANCE or EXIT), rather than a strict character-for-character output requirement for the student's code.

Visual Representation in description.md: This means the most critical place where the choice of symbol matters for universal clarity is indeed the description.md file itself. We definitely want the example mazes there to be clear and render correctly for everyone, regardless of their system or font. I'm completely open to changing the arrow symbol to > (GREATER-THAN SIGN) as you suggested, as it's certainly more robust and less prone to rendering inconsistencies than ⇨.

String Output and Test Runners: For tracks that do require generating a string representation of the maze (like AWK, Bash, and potentially PowerShell), the output will be checked by a test runner. These runners can generally be configured or are run in environments where we have more control over character interpretation. For context, the current Mazy Mice exercise on the AWK track does use the ⇨ symbol, and the tests there are designed to check for it. This shows it can work, but I understand this doesn't guarantee universal rendering fidelity in all contexts, especially in diverse browser/editor views of the markdown.


I'm happy to update the description.md in the problem specification to use > as the character for entrance/exit points. This would ensure the visual examples are solid for all tracks and that the "valid characters" (as mentioned in the canonical-data.json comments) are reliably represented.

Since this change would affect the canonical problem specification used by all tracks, I think it's a good idea to also solicit opinions from other colleagues involved in problem specifications or track maintenance to ensure we're all on board with this (or if there are other preferences/considerations).

What are your thoughts on this approach? I appreciate your input on this!

IsaacG replied:
If the exercise is focused on maze generation absent of the drawing (ie cells having type `{WALL, PATH, ENTRANCE, EXIT}`), I'd be in favor of simplifying the bash and awk exercises to remove the complexity of having more than 4 output chars.

Commit:
Add introduction for Mazy Mice exercise

Introduce Mickey and Minerva, the main characters of the Mazy Mice exercise. The introduction sets the stage for the maze-solving challenge, emphasizing a single correct path to guide users.

Commit:
Simplify and restructure maze instructions.

Moved content from description to instructions for clarity and streamlined examples section. Removed redundant box-drawing character table while keeping key details for maze generation intact.

rabestro commented on exercises/mazy-mice/description.md:
```diff
@@ -0,0 +1,77 @@
+
+Meet Mickey and Minerva, two clever mice who love to navigate their way through a maze to find cheese. They enjoy a good challenge, but with only their tiny mouse brains, they prefer if there is only one correct path to the cheese.
+
+## Instructions
```
done

Commit:
Remove unused link reference from instructions

The link reference "[box-drawing]" was not used anywhere in the document, so it has been removed to clean up the file. This change improves clarity and eliminates unnecessary clutter.

ErikSchierboom commented on exercises/mazy-mice/introduction.md:
```diff
@@ -0,0 +1,3 @@
+# Introduction
+
+Meet Mickey and Minerva, two clever mice who love to navigate their way through a maze to find cheese. They enjoy a good challenge, but with only their tiny mouse brains, they prefer if there is only one correct path to the cheese.
```
Tiny nit (we use one line per sentence):

```suggestion
Meet Mickey and Minerva, two clever mice who love to navigate their way through a maze to find cheese.
They enjoy a good challenge, but with only their tiny mouse brains, they prefer if there is only one correct path to the cheese.
```

ErikSchierboom commented on exercises/mazy-mice/instructions.md:
```diff
@@ -58,20 +53,5 @@ The rectangular maze 6x18 cells
-|     ⇨     | rightwards white arrow                 | U+21E8  |
-
 [recursive-backtracker]: https://en.wikipedia.org/wiki/Maze_generation_algorithm
 [box-drawing]: https://en.wikipedia.org/wiki/Box-drawing_character
```
This is no longer needed.

```suggestion
```

rabestro replied:
Fixed

rabestro committed:
Update exercises/mazy-mice/introduction.md

Co-authored-by: Erik Schierboom <erik_schierboom@hotmail.com>

senekor commented on exercises/mazy-mice/instructions.md:
```diff
@@ -0,0 +1,56 @@
+    │ └─┐ ┌─┐ │ │ │ ├── ├───┐ │ │ ──┼── │
+    │   │ │ │   │   │   │   │ │ │   │   │
+    └── │ │ ├───┴───┤ ┌─┘ ┌─┘ │ ├── │ ──┤
+    ⇨   │   │       │ │   │   │ │   │   │
```
There are still some wide arrows in the documentation.

rabestro replied:
> There are still some wide arrows in the documentation.

Yes, if I understood correctly we can keep UTF-8 symbols in description for information (illustration) purposes. The illustration rendered correctly for this exercise (AWK and Java tracks). I've deleted the table with symbols as mentioned by @ErikSchierboom

senekor replied:
I guess you mean it's rendered correctly on the website? That's not the only place where people will see it. It's rendered badly here on GitHub and it will be rendered badly in all the various editors people use locally. It's kind of a minor thing but I don't see the benefit of using a fancy arrow symbol in ascii art instead of just `>`, `->` or `=>`.

tasxatzial replied:
in the past i've used this one `➜` in some of my PRs and it seems to be rendered fine in editors and the website.

senekor commented:
Some general feelings of mine about this exercise: I don't like exercises with randomness. They are hard to test properly. I think this can be seen here pretty well, where the test data is not very useful. Basically the only information in every test case is the description of what property should be checked. Like Erik said, you can't use that to generate the actual test cases. Every track will have to write the test cases from scratch.

I also have a feeling that it's pretty easy to generate a maze that technically complies with the properties but would be very boring to solve.

I have a very opinionated idea: We could force users to implement a specific algorithm to generate the maze. Then we know what kind of "random" decision the algorithm will have to make at every step. We can supply the list of decisions of the algorithm as input and the precise maze that must be generated as data to test against.

cons:
- Users have no freedom to choose the maze generation algorithm they want.

pros:
- Users have no freedom to choose a boring maze generation algorithm, i.e. "cheesing" the exercise.
- We can create _much_ better tests that lend themselves to TTD like our other exercises. We can start off with simple mazes and work up to mazes with more complicated edge cases.
- The tests can be much better suited for test generation, meaning lower effort to implement it on each track.

ErikSchierboom commented:
@senekor Would the maze generation algorithm be easy to explain in steps?

senekor commented:
You mean easy to explain to students? I guess it depends on the algorithm we choose. There's a list [here](https://en.wikipedia.org/wiki/Maze_generation_algorithm). I think it doesn't matter much if we choose an easy or a difficult one, in either case the difficulty is at least standardized. This will also make it easier for tracks to select an appropriate difficulty setting.

ErikSchierboom commented:
I'm 's bit torn. I quite like the randomness aspect as that fits my mental model or generating mazes. That said, the implementation of its tests is a lot harder. Maybe it would be useful if someone tried implementing this exercise in their language of choice.

jiegillet commented:
[I don't know how useful this will be, I have not read the PR and I'm only reacting to the last comment, but] we have a concept exercise in the Elm track that's about generating a random maze.
[Link to the exercise](https://exercism.org/tracks/elm/exercises/maze-maker/), [link to my solution](https://exercism.org/tracks/elm/exercises/maze-maker/solutions/jiegillet), [link to github](https://github.com/exercism/elm/tree/main/exercises/concept/maze-maker).
Elm has this cool thing where you can do fuzz tests and check the distribution of the generated solutions (how many solutions have such or such property).

senekor commented:
I think it would be pretty easy to make most tests deterministic and then have one or more random ones at the end. Those wouldn't be primarily for checking correctness, but just to drive home the feeling of generating mazes randomly. That would even teach the additional valuable lesson that you can split programs with randomness in a deterministic part that can still be well tested. Basically the "functional core, imperative shell" concept.

ErikSchierboom commented:
I'd like to get some more people's thoughts on this. @exercism/reviewers would you mind chiming in here?

IsaacG commented:
> I'd like to get some more people's thoughts on this. @exercism/reviewers would you mind chiming in here?

Should this be on the forum and targeted at maintainers, who would be the ones building the tests?

Is the concern about the complexity and maintenance around keeping tests reliable? The cost of needing to write a maze solver to test this exercise?

Glenn and I have already approved this exercise in the `awk` track which is at least somewhat of an implicit approval, though we only approved the test runner and didn't have to write it ourselves.

On the one hand, it's a fun and interesting exercise for students to solve. On the other hand, writing the tests for it would definitely be harder than most (all?) of the other exercises by a large margin. Is that an issue if maintainers can opt to omit this exercise? Would clear docs with guidance on writing tests help?

kotp commented:
> > Should this be on the forum and targeted at maintainers, who would be the ones building the tests?
> 
> I'll open a forum post. Sorry for the delay @rabestro!

The forum post can be found Mazy Mice exercise proposal [here][1].

[1]: http://forum.exercism.org/t/new-exercise-proposal-mazy-mice/17277/9
